tests/models/audio_spectrogram_transformer/test_feature_extraction_audio_spectrogram_transformer.py

```diff
@@ -172,3 +174,136 @@ def test_integration(self):
         input_values = feature_extractor(input_speech, return_tensors="pt").input_values
         self.assertEquals(input_values.shape, (1, 1024, 128))
         self.assertTrue(torch.allclose(input_values[0, 0, :30], EXPECTED_INPUT_VALUES, atol=1e-4))
+
+    def test_feat_extract_from_and_save_pretrained(self):
+        feat_extract_first = self.feature_extraction_class(**self.feat_extract_dict)
+
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            saved_file = feat_extract_first.save_pretrained(tmpdirname)[0]
+            check_json_file_has_correct_format(saved_file)
+            feat_extract_second = self.feature_extraction_class.from_pretrained(tmpdirname)
+
+        dict_first = feat_extract_first.to_dict()
+        dict_second = feat_extract_second.to_dict()
+        self.assertDictEqual(dict_first, dict_second)
+
+    def test_feat_extract_to_json_file(self):
+        feat_extract_first = self.feature_extraction_class(**self.feat_extract_dict)
+
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            json_file_path = os.path.join(tmpdirname, "feat_extract.json")
+            feat_extract_first.to_json_file(json_file_path)
+            feat_extract_second = self.feature_extraction_class.from_json_file(json_file_path)
+
+        dict_first = feat_extract_first.to_dict()
+        dict_second = feat_extract_second.to_dict()
+        self.assertEqual(dict_first, dict_second)
+
+
+@require_torch
+@unittest.mock.patch(
+    "transformers.models.audio_spectrogram_transformer.feature_extraction_audio_spectrogram_transformer.is_speech_available",
+    lambda: False,
+)
+class ASTFeatureExtractionWithoutTorchaudioTest(SequenceFeatureExtractionTestMixin, unittest.TestCase):
```

Comment:
Very nice. I think you can either add `# Copied from` on top of some tests, or (not sure if it's possible) find a way to use `parametrized` to mock the absence of the package as a parameter to avoid code duplications. Not very important `# Copied from` will be great

Reply:
Hello @ylacombe ! There is something like `unittest.mock.Mock()` but I never used it (yet) myself.

Search in `tests/models/t5/test_modeling_t5.py` 

```
    def test_fp16_fp32_conversion(self):
        r"""
        A test to check whether the argument `keep_in_fp32_modules` correctly does its job
        """
        orig_import = __import__
        accelerate_mock = unittest.mock.Mock()

        # mock import of accelerate
        def import_accelerate_mock(name, *args, **kwargs):
            if name == "accelerate":
                if accelerate_available:
                    return accelerate_mock
                else:
                    raise ImportError
            return orig_import(name, *args, **kwargs)
```
and let me know how you feel. I can take a look too (good to learn anyway)

Reply:
Sorry, I missed the above comment. So here is 2 classes instead of 2 test methods.

Is the code in the new `ASTFeatureExtractionWithoutTorchaudioTest` be identical to the original `ASTFeatureExtractionTest`? If so, maybe try make it a subclass of `ASTFeatureExtractionTest` but decorated with `unittest.mock.patch` or something similar?

Reply:
Thanks for this!

Making a subclass works for me, I'll try it